File: pythonr-pipeline/pythonr-pipeline/python-part/xgb-deployment/deployment.py
"""
The file containing the deployment code is required to be called 'deployment.py' and should contain the 'Deployment'
class and 'request' method.
"""

import pandas as pd
import numpy as np
import os
from joblib import load

class Deployment:

    def __init__(self, base_directory, context):
        """
        Initialisation method for the deployment. It can for example be used for loading modules that have to be kept in
        memory or setting up connections. Load your external model files (such as pickles or .h5 files) here.

        :param str base_directory: absolute path to the directory where the deployment.py file is located
        :param dict context: a dictionary containing details of the deployment that might be useful in your code.
            It contains the following keys:
                - deployment (str): name of the deployment
                - version (str): name of the version
                - input_type (str): deployment input type, either 'structured' or 'plain'
                - output_type (str): deployment output type, either 'structured' or 'plain'
                - environment (str): the environment in which the deployment is running
                - environment_variables (str): the custom environment variables configured for the deployment.
                    You can also access those as normal environment variables via os.environ
        """

        print("Initialising xgboost model")

        XGBOOST_MODEL = os.path.join(base_directory, "xgb-deployment.joblib")
        self.model = load(XGBOOST_MODEL)

    def request(self, data):
        """
        Method for deployment requests, called separately for each individual request.

        :param dict/str data: request input data. In case of deployments with structured data, a Python dictionary
            with as keys the input fields as defined upon deployment creation via the platform. In case of a deployment
            with plain input, it is a string.
        :return dict/str: request output. In case of deployments with structured output data, a Python dictionary
            with as keys the output fields as defined upon deployment creation via the platform. In case of a deployment
            with plain output, it is a string. In this example, a dictionary with the key: output.
        """
        print('Loading data')
        input_data = pd.read_csv(data['clean_data'])
        
        print("Prediction being made")
        prediction = self.model.predict(input_data.values)
        
        # Writing the prediction to a csv for further use
        print('Writing prediction to csv')
        pd.DataFrame(prediction).to_csv('prediction.csv', header = ['house_prices'], index_label= 'index')
        
        return {
            "prediction": 'prediction.csv'
        }

File: pythonr-pipeline/pythonr-pipeline/r-part/dea-deployment/deployment.R
#' @title Init
#' @description Initialisation method for the deployment.
#'     It can for example be used for loading modules that have to be kept in memory or setting up connections.

init <- function(base_directory, context) {
  print("Initialising My Deployment")
}

#' @title Request
#' @description Method for deployment requests, called separately for each individual request.
#' @return output data (str or named list) request result
#'     - In case of structured output: a named list, with as keys the output fields as defined upon deployment creation
#'     - In case of plain output: a string
request <- function(input_data, base_directory, context) {
 print("Processing request for My Deployment")

 # Load in the required libraries
 library(ggplot2)
 library(GGally)
 library(gridExtra)
 library(scales)
  
 df <- read.csv(input_data[["raw_data"]], stringsAsFactors = FALSE)
 # Remove the date column
 df <- subset(df, select = -c(date))


 condition <- df$condition
 grade <- df$grade
 price <- df$price
  
 print(paste("rows:", nrow(df), "cols:", ncol(df)))

 # Lets look at the distribution of house condtion, grade and price:
 p1 <- qplot(condition, data = df, geom = "bar",
              main = "Number of houses by condition")

 p2  <- qplot(grade, data = df, geom = "bar",
              main = "Number of houses by grade")

 p3 <- ggplot(df, aes(x = price)) + geom_density(stat = "density") +
    xlab("price")  + ggtitle("Price distribution") 


  
 # Look at price (log10) vs other features:
 ggplot(df, aes(x = log10(price), y = sqft_living)) +
  geom_smooth() +
  scale_y_continuous(labels = comma) +
  scale_x_continuous(labels = comma) +
  ylab("sqft of living area") +
  geom_point(shape = 1, alpha = 1/10) +
  ggtitle("Price (log10) vs sqft of living area")

 ggplot(df, aes(x = grade, y = log10(price))) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  coord_flip() +
  geom_point(shape = 1, alpha = 1/10) +
  ggtitle("Price (log10) vs grade")
  

 ggplot(df, aes(x = condition, y = log10(price))) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  coord_flip() +
  geom_point(shape = 1, alpha = 1/10) +
  ggtitle("Price (log10) vs condition")
  

 ggplot(df, aes(x = as.factor(floors), y = log10(price))) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  xlab("floors") +
  coord_flip() +
  geom_point(shape = 1, alpha = 1/10) +
  ggtitle("Price (log10) vs number of floors")
 
 
 # Look at how the different features correlate
  ggcorr(df, hjust = 0.8, layout.exp = 1) +
  ggtitle("Correlation between house features")
    

 # Create a DataFrame that only contains the columns which we will use to predict the house prices
 data <- df[, c('sqft_living', 'grade', 'sqft_above', 'sqft_living15', 'bathrooms', 'view', 'sqft_basement', 'lat', 'waterfront', 'yr_built', 'bedrooms')]

 output_path <- file.path(getwd(), "data.csv")
 write.csv(data, output_path, row.names = FALSE)
  
 return(list("clean_data" = output_path))
  
}
head(data)

